Collections backend

BenHenning opened the pull request:
Provides all necessary backend functionality to support collections in Oppia.

seanlip commented on core/domain/collection_services.py:
```diff
@@ -0,0 +1,872 @@
+            return None
+
+
+def get_collection_summary_by_id(collection_id):
```
> [@seanlip] Recording offline discussion: let's keep the summaries separate for now, but add a TODO to merge things (and perhaps generate them with a batch job) if there is no divergence by Jan 2016 (say). Also, file an issue for this. We may also want to recharacterize summaries as SearchSnippets or similar, since that's really what they are.

> > [@BenHenning] I may not remember correctly, but I do not think the offline discussion was particularly picky about keeping the summaries separate. Since the changes in rights_manager introduced the concept of an activity (and related activity types for collections and explorations), I'm inclined to unify the two summaries into an ActivitySummary and differentiate between the two summaries based on their activity type (though this might not even be needed since the summaries are differentiated based on their search indexes, anyway).

SGTM.

BenHenning replied:
SGTM!

seanlip commented on core/domain/collection_services.py:
```diff
@@ -0,0 +1,872 @@
+    logging.info('Collection with id %s was loaded.' % collection_id)
+
+
+# TODO(bhenning): Cleanup search logic and abstract it between explorations and
```
Definitely +1 to unifying the search functionality for explorations and collections into a separate search_services (not activity_services) file. I don't think I have strong feelings with regards to having one or two indexes. (We could even postpone the decision and have three -- 'explorations', 'collections', 'all'.)

> If I understand this suggestion correctly, we would instead have a search_manager which has an actor class for the index and all the search functionality which acts upon the index. Then, the search functions in both collection_services and exp_services would just use that manager with their respective indexes. The call sites for exploration search functions wouldn't change. The unified ActvitySummary would still live in activity_domain.py.

Not quite, I think -- the index wouldn't literally have or be an 'actor', it's just the design pattern of the Actor class that I was referring to. So there would be a class like this one:

    class SearchIndex(object):
        """Domain object for a search index."""
        def __init__(self, name):
            self.name = name

        def ....

BenHenning replied:
As we discussed offline, we want to avoid tightly coupling search functionality between explorations and collections so this is not going to be done as part of this branch.

seanlip commented on core/domain/collection_domain.py:
```diff
@@ -0,0 +1,559 @@
+
+    def delete_node(self, exploration_id):
+        node_index = self._find_node(exploration_id)
+        if node_index == None:
```
if node_index is None

BenHenning replied:
Fixed.

seanlip commented on core/domain/subscription_services.py:
```diff
@@ -57,14 +57,14 @@ def subscribe_to_activity(user_id, activity_id):
 
 def get_activity_ids_subscribed_to(user_id):
-    """Returns a list with all subscriptions' ids of given user.
+    """Returns a list with ids of all activities (i.e., explorations and
```
Is this just 'explorations'? (See [old comment](https://code.google.com/p/oppia/source/diff?spec=svn5c2606468772bf7249a41f3d65d38443c5de2b39&name=collections-backend&r=5c2606468772bf7249a41f3d65d38443c5de2b39&format=side&path=/core/domain/subscription_services.py) for get_collection_ids_subscribed_to() -- just marking this as needing to be resolved somehow.

BenHenning replied:
Yes, the comment is wrong unless we want to update subscription_services to actually store activities, rather than explorations and collections individually. I think for the sake of time, we might as well just store them separately. In the future, we might want to consider unifying them and performing a migration. Also, fixed the comment. I updated all references to 'activity' to instead refer to 'exploration' in the cases where it did not refer to both a collection and exploration.

seanlip commented on core/domain/rights_manager.py:
```diff
@@ -215,6 +243,52 @@ def is_exploration_cloned(exploration_id):
+            return get_collection_rights(activity_id)
+    except Exception:
+        pass
+    return None
```
> @seanlip: This can go in the previous line?
>> @BenHenning: Except in the invalid case when activity_type is neither ACTIVITY_TYPE_EXPLORATION nor ACTIVITY_TYPE_COLLECTION. It would technically return None anyway, but that seems a bit dirty to me. What do you think?

Oh, I see. OK, probably fine to leave as-is. Though, why don't we just throw an error if it's not either of the given types?

BenHenning replied:
Sgtm. Updated to throw an exception.

seanlip commented:
OK, I've made all the residual comments -- no more to add. Thanks @BenHenning! Also, the following is the list of things I'm aware of that's still TODO:

- [ ] tests for rights management for collections (do you still need my input here)?
- [ ] modify user_jobs.RecentUpdatesMRJobManager to take subscribed collections into account (and add tests)
- [ ] ensure that collection_services.get_collection_summaries_matching_query is tested somehow
- [ ] introduce search_services file
- [ ] consolidate activity summaries

BenHenning commented:
@seanlip 

Made various changes and also made sure the branch was caught up to develop. As discussed offline, I am not consolidating the activity summaries or introducing a common search services file since we are not sure which aspects of collections and exploration search functionality will be shared and we want to avoid tightly coupling those services right now.

I did not introduce a test for collection_services.get_collection_summaries_matching_query simply because the respective one in exp_services is not tested and I wanted to see whether you think they both should have tests added for them. Other aspects of the summaries and search functionality is tested for explorations. Please let me know your thoughts overall on the testing coverage of collection_services.

Thanks!

seanlip commented on core/domain/subscription_services_test.py:
```diff
@@ -162,55 +171,150 @@ def test_creating_exploration_results_in_subscription(self):
+
+        self.assertEqual(
+            self._get_collection_ids_subscribed_to(self.owner_id),
+            [COLLECTION_ID])
```
Test that if someone creates an exploration and it's later included in a collection, this doesn't result in a subscription

Test that if someone creates a collection and includes someone else's exploration in it, they don't get subscribed to that exploration (or should they be?)

BenHenning replied:
Both are tested in a new test. I think there's good argument for both sides of whether a collection author should be subscribed to an exploration he/she includes. I'm leaning to not subscribing them for now since that's the path of least resistance; we can add it later if it seems to be a desired feature from collection authors.

seanlip commented on core/domain/user_jobs.py:
```diff
@@ -36,7 +40,8 @@ class DashboardRecentUpdatesAggregator(jobs.BaseContinuousComputationManager):
     of explorations and feedback threads to show on a user's dashboard.
 
-    This job does not have a realtime component. There will be a delay in
+    This job does not have a working realtime component. The
```
suggest s/component. The/component: the

BenHenning replied:
Done.

seanlip commented on core/domain/user_jobs.py:
```diff
@@ -82,56 +87,49 @@ def get_recent_notifications(cls, user_id):
-
-        activity_models = exp_models.ExplorationModel.get_multi(
+    def _get_most_recent_activity_commits(
+            activity_model_cls, activity_ids_list, tracked_models_for_feedback,
```
Can you document? I have no idea what some of these variables represent.

BenHenning replied:
Added thorough documentation for this function.

seanlip commented on core/domain/user_jobs.py:
```diff
@@ -140,10 +138,48 @@ def map(item):
-                        feedback_thread_ids_list.append(thread['thread_id'])
+                tracked_models_for_feedback.append(activity_model)
+
+        return most_recent_commits
```
I incline to return a 2-tuple here in the form of (most_recent_commits, tracked_models_for_feedback), instead of relying on undocumented side-effects to lists that are passed in.

BenHenning replied:
Done.

seanlip commented on core/domain/user_jobs.py:
```diff
@@ -140,10 +138,48 @@ def map(item):
+                if thread['thread_id'] not in feedback_thread_ids_list:
+                    feedback_thread_ids_list.append(thread['thread_id'])
+
+        # TODO(bhenning): Implement a solution to having feedback threasd for
```
nit: threads

BenHenning replied:
Fixed.

seanlip commented on core/domain/user_jobs.py:
```diff
@@ -233,6 +269,40 @@ def map(item):
                                 'id': item.id
                             })
+        elif isinstance(item, collection_models.CollectionRightsModel):
+            if item.deleted:
```
Feels like a lot of this duplicates the previous clause and can be refactored into a separate private method; wdyt?

BenHenning replied:
Because 'yield' cannot happen in a separate private method (at least I couldn't get it to work). Since most of the code is 'yield' it seemed like most of the duplication wouldn't be removed or I would have to introduce some ugly patterns (such as returning objects which are yielded). I don't like the duplication much, but it's cleaner and easier to maintain here.

seanlip replied:
That's reasonable. Could you document the rationale, so that future developers understand why? I think it's non-obvious.

seanlip replied:
(Flagging this.)

BenHenning replied:
Added a note for this.

seanlip commented on core/domain/user_jobs_test.py:
```diff
@@ -460,11 +479,128 @@ def test_subscribing_to_exp_subscribes_to_its_feedback_threads(self):
     EXP_ID_2 = 'exp_id_2'
+    COLLECTION_ID = 'col_id'
+    COLLECTION_ID_2 = 'col_id_2'
+    COLLECTION_EXP_ID = 'col_exp_id'
```
Took a while for me to figure out what this means. Maybe call it EXP_ID_FOR_COLLECTION_1 and s/COLLECTION_ID/COLLECTION_ID_1, s/EXP_ID/EXP_ID_1

BenHenning replied:
Done!

seanlip commented on core/domain/user_jobs_test.py:
```diff
@@ -651,3 +794,105 @@ def test_deleted_exploration(self):
+            sorted(user_a_subscriptions_model.collection_ids),
+            sorted([self.COLLECTION_ID, self.COLLECTION_ID_2]))
+
+    def test_deleted_collection(self):
```
What if you delete the collection but not the exploration, and vice versa?

BenHenning replied:
I'm not so sure this needs to be tested. The test is testing against an empty collection.

Any behavior with subscribing to explorations happens as part of the standard exploration subscriptions pipeline and is guarded by the exploration subscription tests. Collections are independent of this behavior and have been proved in the immediate pipeline not to lead to collections and explorations from different authors crossing subscription. I added one test to guard against similar behavior in the user job.

seanlip replied:
Oops, sorry -- I meant that there should probably be tests that verify the behaviour in the cases I mentioned (my comment wasn't specific to this particular test).

BenHenning replied:
Please let me know if the new tests test the behaviors you were referring to.

seanlip replied:
Yes, this is good. Thanks!

seanlip commented:
This all looks very good. Thanks @BenHenning!

I think the test coverage for collection_services looks very good. The only exception is that I think we should test collection_services.get_collection_summaries_matching_query (and the same for exp_services). If you look at the coverage report, it seems like that's the major gap.

BenHenning commented:
@seanlip Addressed your review comments, thanks for the review! The only thing I did not do was test get_collection_summaries_matching_query because it will be a nearly identical clone to the exp_services version, so I wanted to make sure the exp_services version was well-suited before implementing it.

seanlip commented:
Thanks! Looks good overall, just a few more things.

BenHenning commented:
@seanlip Thanks for the review! I have addressed your comments and implemented the final parts of collections-backend that we discussed. It should be code complete right now as far as I am aware.

BenHenning commented:
@seanlip I addressed all of your review comments, including the one you flagged earlier, but excluding the one about still referring to a commit_type (I apologize but I need clarification on what you were specifically referring to regarding that comment).

With one small change that was not pushed, all but 1 integration test passes (one of the embedding integration tests fail to find an element). Since collections-backend does not affect frontend behavior and all of the critical editor/exploration integration tests pass, I am not going to hinge the merging of this branch into develop on all of the integration tests passing.

seanlip commented:
Thanks @BenHenning, this is great!

I suggest adding a test for non-existent rights objects, as we discussed offline. Once that's done I think this is good to go -- please feel free to merge into develop.

BenHenning commented:
Thanks for the review!
